src/github.com/matrix-org/dendrite/roomserver/api/output.go

```diff
@@ -43,6 +40,14 @@ type OutputRoomEvent struct {
 	// If the LastSentEventID doesn't match what they were expecting it to be
 	// they can use the LatestEventIDs to request the full current state.
 	LastSentEventID string
+	// The state event IDs that are part of the state at the event, but not
+	// part of the current state. Together with the StateBeforeRemovesEventIDs
+	// this can be used to construct the state before the event from the
+	// current state.
+	StateBeforeAddsEventIDs []string
```

Comment:
I'm uncomfortable with this. AIUI, consumers will be incrementally building up the complete room state based on `AddsStateEventIDs` and `RemovesStateEventIDs`. How does `StateBeforeAddsEventIDs` and `StateBeforeRemovesEventIDs` fit into this? You state that:

> this can be used to construct the state before the event from the current state.

But this isn't helpful. My questions are:
 - Why would a consumer ever need to "construct the state before the event from the current state"? I assume this has something to do with routing, but you don't state as much.
 - Should consumers remember the events in this slice, and use it to make the current room state? I imagine not, but you don't say this.
 - What are the clobbering semantics? I assume the order is:
    * Insert `AddsStateEventIDs`
    * Remove `RemovesStateEventIDs`
    * Insert `StateBeforeAddsEventIDs`
    * Remove `StateBeforeRemovesEventIDs`

    But you don't state this.